DynamicModeler/Logic/vtkSlicerDynamicModelerROICutTool.cxx

```diff
@@ -0,0 +1,322 @@
+{
+  return "ROI cut";
+}
+
+//----------------------------------------------------------------------------
+bool vtkSlicerDynamicModelerROICutTool::RunInternal(vtkMRMLDynamicModelerNode* surfaceEditorNode)
+{
+  if (!this->HasRequiredInputs(surfaceEditorNode))
+    {
+    vtkErrorMacro("Invalid number of inputs");
+    return false;
+    }
+
+  vtkMRMLModelNode* outputInsideModelNode = vtkMRMLModelNode::SafeDownCast(surfaceEditorNode->GetNodeReference(ROI_CUT_OUTPUT_INSIDE_MODEL_REFERENCE_ROLE));
+  vtkMRMLModelNode* outputOutsideModelNode = vtkMRMLModelNode::SafeDownCast(surfaceEditorNode->GetNodeReference(ROI_CUT_OUTPUT_OUTSIDE_MODEL_REFERENCE_ROLE));
+  if (!outputInsideModelNode && !outputOutsideModelNode)
+    {
+    // Nothing to output
+    return true;
+    }
+
+  vtkMRMLModelNode* inputModelNode = vtkMRMLModelNode::SafeDownCast(surfaceEditorNode->GetNodeReference(ROI_CUT_INPUT_MODEL_REFERENCE_ROLE));
+  if (!inputModelNode)
+    {
+    vtkErrorMacro("Invalid input model node!");
+    return false;
+    }
+  if (!inputModelNode->GetMesh() || inputModelNode->GetMesh()->GetNumberOfPoints() == 0)
+    {
+    if (outputInsideModelNode && outputInsideModelNode->GetPolyData())
+      {
+      outputInsideModelNode->GetPolyData()->Initialize();
+      }
+    if (outputOutsideModelNode && outputOutsideModelNode->GetPolyData())
+      {
+      outputOutsideModelNode->GetPolyData()->Initialize();
+      }
+    return true;
+    }
+
+  vtkMRMLMarkupsROINode* roiNode = vtkMRMLMarkupsROINode::SafeDownCast(surfaceEditorNode->GetNodeReference(ROI_CUT_INPUT_ROI_REFERENCE_ROLE));
+  if (!roiNode)
+    {
+    if (outputInsideModelNode)
+      {
+      outputInsideModelNode->SetAndObservePolyData(vtkNew<vtkPolyData>());
+      }
+    if (outputOutsideModelNode)
+      {
+      outputOutsideModelNode->SetAndObservePolyData(vtkNew<vtkPolyData>());
+      }
+    return true;
+    }
+
+  vtkNew<vtkPlanes> planes;
+  roiNode->GetTransformedPlanes(planes);
+
+  vtkNew<vtkImplicitBoolean> planeFunction;
+  if (roiNode->GetInsideOut())
+    {
+    // In inside out mode, we want to remove the intersection of all planes inside the ROI.
+    planeFunction->SetOperationTypeToIntersection();
+    }
+  else
+    {
+    // If inside out mode is off, we want to remove any regions that are outside of the ROI.
+    planeFunction->SetOperationTypeToUnion();
+    }
+
+  vtkNew<vtkPlaneCollection> planeCollection;
+  for (int i = 0; i < planes->GetNumberOfPlanes(); ++i)
+    {
+    vtkPlane* currentPlane = planes->GetPlane(i);
+
+    // vtkClipPolyData will remove the inside of the ROI (negative implicit function), so we need to invert the planes so that the correct
+    // region is preserved.
+    double normal[3] = { 1,0,0 };
+    currentPlane->GetNormal(normal);
+    vtkMath::MultiplyScalar(normal, -1.0);
+
+    double origin[3] = { 0.0, 0.0, 0.0 };
+    currentPlane->GetOrigin(origin);
+
+    if (planes->GetTransform())
```

Comment:
Getting bounding plane collection is not trivial, especially if the ROI is under non-linear transform. Instead of implementing getting planes here, we should have a method for this in the ROI node. The method should be able to return the plane collection in local or world coordinates, and should have an inside-out option.

Reply:
I've updated the code to use the new GetPlanesWorld function.